tools/run-csx.ps1

```diff
@@ -0,0 +1,96 @@
+# Treat all errors as terminating
+$ErrorActionPreference = "Stop"
+
+$ToolsDir = $PSScriptRoot
+$RepoRoot = "$ToolsDir\.."
+$NuGetVersion = "4.1.0"
+$NuGetCacheDir = "$($Env:LOCALAPPDATA)\.nuget\v$NuGetVersion"
+$NuGetCacheFile = "$NuGetCacheDir\NuGet.exe"
+$NuGetDir = "$ToolsDir\.nuget"
+$NuGetExe = "$NuGetDir\NuGet.exe"
+$PackagesDir = "$ToolsDir\packages"
+$PackagesConfig = "$ToolsDir\packages.config"
+$PackagesHashFile = "$ToolsDir\packages.config.sha1"
+$CsiExe = "$PackagesDir\Microsoft.Net.Compilers\tools\csi.exe"
+
+function EnsureDirectoryExists($path)
+{
+    if (!(Test-Path $path -PathType Container)) {
+        New-Item -ItemType Directory  $path | Out-Null
+    }
+}
+
+function DownloadNuGetToCache()
+{
+    $NuGetUrl = "https://dist.nuget.org/win-x86-commandline/v$NuGetVersion/NuGet.exe"
+    if (!(Test-Path $NuGetCacheFile)) {
+        EnsureDirectoryExists($NuGetCacheDir)
+        Write-Host "Downloading '$NuGetUrl' to '$NuGetCacheFile'..."
+        Invoke-WebRequest $NuGetUrl -OutFile $NuGetCacheFile
+    }
+}
+
+function CopyNuGetLocally()
+{
+    EnsureDirectoryExists($NuGetDir)
+    Copy-Item $NuGetCacheFile $NuGetExe
+}
+
+function ReadPreviousPackagesHash()
+{
+    if (Test-Path $PackagesHashFile)
+    {
+        (Get-Content $PackagesHashFile).Trim()
+    }
+    else
+    {
+        ""
+    }
+}
+
+function WritePackagesHash($hash)
+{
+    Set-Content $PackagesHashFile $hash
+}
+
+function ComputePackagesHash()
+{
+    (Get-FileHash -Algorithm SHA1 -Path $PackagesConfig).Hash
+}
+
+function ClearPackagesDir()
+{
+    Write-Host "$PackagesConfig has changed, clearing packages to force reinstall..."
+    Remove-Item -Recurse -Force $PackagesDir
+}
+
+function ClearPackagesIfChanged()
+{
+    # check if packages.config have changed
+    $PreviousPackagesHash = ReadPreviousPackagesHash
+    $PackagesHash = ComputePackagesHash
+    if ($PackagesHash -ne $PreviousPackagesHash) {
+        # if packages.config has changed, clear packages and store new hash
+        ClearPackagesDir
+        WritePackagesHash($PackagesHash)
+    }
+}
+
+function RestorePackages()
+{
+    & $NuGetExe install $PackagesConfig -OutputDirectory $PackagesDir -Verbosity quiet -ExcludeVersion
+}
+
+# change working directory to the repository root
+Push-Location $RepoRoot
+
+DownloadNuGetToCache
+CopyNuGetLocally

```

Comment:
Yeah, I stumble over this as well.  I think @adamralph originated the behaviour or imported from somewhere else.
I wouldn't mind if either the caching or the copying went away (I don't mind the build script writing to the AppData folder, but don't mind it not doing it either).
The current system does give a way to always have the correct nuget in the local directory (because the cache directory is versioned) without downloading it all the time or having the rest of the build system need to know what the version is.
I guess it also allows one to switch between (already cached) versions while disconnected. but that's a small concern.

Reply:
I don't even mind the idea of _committing_ the nuget exe, if we want.

Reply:
You might find it simpler to do away with scripting completely, as I've done with xbehave and @blairconrad has done with SelfInitializingFakes. It allows you to drop nuget.exe, csi.exe, packages.config, etc. All you need is a [csproj](https://github.com/xbehave/xbehave.net/blob/10c5b8110adedea7fc50ebaa809121bc367cc87e/targets/Targets.csproj) for package references and a [Program.Main](https://github.com/xbehave/xbehave.net/blob/10c5b8110adedea7fc50ebaa809121bc367cc87e/targets/Program.cs#L9) for your build "script". You can then invoke simply with [dotnet run](https://github.com/xbehave/xbehave.net/blob/10c5b8110adedea7fc50ebaa809121bc367cc87e/build.cmd#L4).

Reply:
Yes, it might make things easier...

Reply:
I am enjoying the compiled `Targets` in SelfInitializingFakes.
However, IIRC, we were keeping nuget.exe around because we needed it to package tha Analyzer. My memory is weak, but we might not have been able to create the meta-package that includes both C# and VB. Which maybe we should drop anyhow…

Reply:
> Which maybe we should drop anyhow…

👍 I can't see any use for that package. The number of mixed C# and VB solutions is probably vanishingly small.

Reply:
Actually, a mixed VB/C# solution isn't a use-case either, since an analyzer is installed in a _project_, which can't be mixed language.

Reply:
No, but the meta-package does provide a convenient way to update both C# and VB versions at once. For me, it's not enough of an advantage to continue with it, though.

Reply:
Another option is to use [`dotnet script`](https://github.com/filipw/dotnet-script) instead of manually invoking csi.exe. It [natively supports NuGet references](https://github.com/filipw/dotnet-script#nuget-packages-1). It requires .NET Core SDK 2.1+, though, since it's a global tool.

Reply:
That's a good point!